subsys/net/ip/dhcpv4.c

```diff
@@ -396,130 +415,123 @@ static void dhcpv4_enter_selecting(struct net_if *iface)
-		break;
-	case NET_DHCPV4_BOUND:
-		iface->config.dhcpv4.state = NET_DHCPV4_RENEWING;
-		NET_DBG("enter state=%s",
-			net_dhcpv4_state_name(iface->config.dhcpv4.state));
-		iface->config.dhcpv4.attempts = 0;
-		dhcpv4_send_request(iface);
-		break;
-	}
+	return true;
 }
 
-static void dhcpv4_t2_timeout(struct k_work *work)
+static void dhcpv4_enter_requesting(struct net_if *iface)
 {
-	struct net_if *iface = CONTAINER_OF(work, struct net_if,
-					    config.dhcpv4.t2_timer);
-
-	NET_DBG("");
+	iface->config.dhcpv4.attempts = 0;
+	iface->config.dhcpv4.state = NET_DHCPV4_REQUESTING;
+	NET_DBG("enter state=%s",
+		net_dhcpv4_state_name(iface->config.dhcpv4.state));
 
-	switch (iface->config.dhcpv4.state) {
-	case NET_DHCPV4_DISABLED:
-	case NET_DHCPV4_INIT:
-	case NET_DHCPV4_SELECTING:
-	case NET_DHCPV4_REQUESTING:
-	case NET_DHCPV4_REBINDING:
-		NET_ASSERT_INFO(0, "Invalid state %s",
-			net_dhcpv4_state_name(iface->config.dhcpv4.state));
-		break;
-	case NET_DHCPV4_BOUND:
-		/* If renewal time and rebinding time are
-		 * misconfigured we may end up with T2 firing before
-		 * T1.  Deal with it as though we had transitioned
-		 * through RENEWAL already.
-		 */
-	case NET_DHCPV4_RENEWING:
-		iface->config.dhcpv4.state = NET_DHCPV4_REBINDING;
-		NET_DBG("enter state=%s",
-			net_dhcpv4_state_name(iface->config.dhcpv4.state));
-		iface->config.dhcpv4.attempts = 0;
-		dhcpv4_send_request(iface);
-		break;
-	}
+	dhcpv4_update_timeout_work(dhcpv4_send_request(iface));
 }
 
 static void dhcpv4_enter_bound(struct net_if *iface)
 {
 	u32_t renewal_time;
 	u32_t rebinding_time;
 
-	k_delayed_work_cancel(&iface->config.dhcpv4.timer);
-
 	renewal_time = iface->config.dhcpv4.renewal_time;
 	if (!renewal_time) {
 		/* The default renewal time rfc2131 4.4.5 */
 		renewal_time = iface->config.dhcpv4.lease_time / 2;
+		iface->config.dhcpv4.renewal_time = renewal_time;
 	}
 
 	rebinding_time = iface->config.dhcpv4.rebinding_time;
 	if (!rebinding_time) {
 		/* The default rebinding time rfc2131 4.4.5 */
 		rebinding_time = iface->config.dhcpv4.lease_time * 875 / 1000;
+		iface->config.dhcpv4.rebinding_time = rebinding_time;
 	}
 
 	iface->config.dhcpv4.state = NET_DHCPV4_BOUND;
 	NET_DBG("enter state=%s renewal=%us rebinding=%us",
 		net_dhcpv4_state_name(iface->config.dhcpv4.state),
 		renewal_time, rebinding_time);
 
-	/* Start renewal time */
-	k_delayed_work_submit(&iface->config.dhcpv4.t1_timer,
-			      K_SECONDS(renewal_time));
+	iface->config.dhcpv4.timer_start = k_uptime_get();
+	iface->config.dhcpv4.request_time = min(renewal_time, rebinding_time);
 
-	/* Start rebinding time */
-	k_delayed_work_submit(&iface->config.dhcpv4.t2_timer,
-			      K_SECONDS(rebinding_time));
+	dhcpv4_update_timeout_work(iface->config.dhcpv4.request_time);
 }
 
-static void dhcpv4_timeout(struct k_work *work)
+#define IFACE_CONTAINER_OF(_dhcpv4)					\
+	CONTAINER_OF(CONTAINER_OF(_dhcpv4, struct net_if_config, dhcpv4), \
```

Comment:
This looks nasty. Why not the original way i.e., using just one CONTAINER_OF as that looks much cleaner?

Reply:
here you can't use one container of: I need to get the net_if, but I have the net_if_dhcpv4, which is insinde net_if_config.